### PR TITLE
fix isinteger

### LINKED ISSUE #120: isinteger(::Normed) incorrect
The code currently assumes that the denominator is `2^f`, which is not the case for `Normed`. e.g.
```
julia> isinteger(Normed{UInt8,7}(1))
false
```

I believe you need to change
```julia
isinteger(x::FixedPoint{T,f}) where {T,f} = (x.i&(1<<f-1)) == 0
```
to
```julia
isinteger(x::Fixed{T,f}) where {T,f} = (x.i&(1<<f-1)) == 0
isinteger(x::Normed{T,f}) where {T,f} = (x.i%(1<<f-1)) == 0
```

### PR DESCRIPTION
**Pull request by johnnychen94**
fixes #120 : `isinteger(1N0f8)` should be true

P.S. I don't quite understand the magic behind bits operation, but the tests pass with this patch.

cc: @cjdoris

### CODE REVIEW

File: test/fixed.jl [hunk 1]
```diff
@@ -146,6 +146,14 @@ end
     @test floatmin(Q11f4) == Q11f4(0.06)
 end
 
+@testset "isinteger" begin
+    # issue #120
+    for T in (Fixed{Int8,6}, Fixed{Int16,8}, Fixed{Int16,10}, Fixed{Int32,16})
+        a = 1
+        @test all(isinteger.(T.(a)))
```
Comment (timholy) [inline test/fixed.jl]: `a` is a scalar here so I don't think you need `all` or the dots.

How about a loop `for a = 0:floor(Int, typemax(T))`? (Here and below.) Or igerf you want to check ones that are non-integers, you could do
```julia
for a = reinterpret(typemin(T)):reinterpret(typemax(T))
    x = reinterpret(T, a)
    @test isinteger(x) == (x == round(Int, x))
end
```
This kind of exhaustive check probably only makes sense for 16bits or less, and probably best put inside a function so it gets compiled.

**Comment by johnnychen94** [comment]
After solving the conflicts, I failed to rebase the commits (not a git expert). You could either squash it or rebase it into two commits (if it's possible):

* make generate_fixedpoint_types a test utils
* fix isinteger

**Comment by johnnychen94** [comment]
Wow, this patch fails in Windows x86. That's really out of my ability then. Can someone take it over?

**Comment by johnnychen94** [comment]
Thank you for putting your time investigating it! I'll find time to fix them.

**Comment by kimikage** [comment]
As I did not care about this issue until yesterday, please take your time.

File: src/normed.jl [hunk 2]
```diff
@@ -169,3 +169,6 @@ end
 
 _unsafe_trunc(::Type{T}, x::Integer) where {T} = x % T
 _unsafe_trunc(::Type{T}, x) where {T}          = unsafe_trunc(T, x)
+
+# predicates
+isinteger(x::Normed{T,f}) where {T,f} = (x.i%(1<<f-1)) == 0
```
[the remark this answers is not in the document]
Comment (kimikage) [reply src/normed.jl]: The above is analogous to `isinteger(x::Fixed)`. Of course, you can use `rawone(Normed{T,f})`.

**Comment by kimikage** [comment]
I am sorry for hindering the `rebase`, but now we can use `isinteger(x) == isinteger(float(x))` to test `Normed{UInt8}`, `Normed{UInt16}` and `Fixed`. 
However, `float(x::Normed{UInt32})` and `float(x::Normed{UInt64})` still have 1-lsb errors. If you want to introduce the random testing for them, extra checks are required.
IMO, the implementation of `isinteger` should be so *primitive* that only the numbers around the boundaries which might possibly cause overflow/underflow should be checked.

**Comment by kimikage** [comment]
I'm ready to fix `isinteger` with minimal tests (i.e. without random tests). 
https://github.com/kimikage/FixedPointNumbers.jl/compare/commonize1...kimikage:isinteger

Since the master has left this branch, I'm going to submit a new PR.

**Comment by johnnychen94** [comment]
@kimikage Please feel free to do so 👍 , recently I just spent too much time here and I'm trying to catch up with my daily research projects in ECNU.

**Pull request closed without merging**